### ISSUE TITLE
ZIO 2.0 error: A Tag may not contain an intersection type, yet have provided

### ISSUE THREAD

**Issue by alphaho** [zio/zio#6643]
When I try to migrate from ZIO 1.0.13 to **ZIO 2.0.0-RC5**, I've got the error message `A Tag may not contain an intersection type, yet have provided`.

This happens  when I update our ZLayer construction from something like:
```scala
val layer: ZLayer[String with Int, Nothing, Foo] = (new Bar(_, _)).toLayer
```

to:
```scala
val layer: ZLayer[String with Int, Nothing, Foo] = ZLayer.fromFunction(new Bar(_, _))
```

The reason we use this style instead of following the Service Pattern 2.0(which would mark `Bar` as case class) is because we want to separate the service classes from the case classes. And we also want to make the type explicit as well as follow the rule to have type annotation on every public member.

---
Playing further with it, I have found the problem may be workaround by either(as attached below):
- drop the type annotation, as shown in `layer2`, `layer6`, `layer7`
- use the recommended `ZLayer.fromFunction(CaseClass.apply _)` with or without the type annotation, as shown in `layer5`, and `layer8`

Beyond that, one tricky finding I have is with `layer4`, `ZLayer.fromFunction(CaseClassBar(_, _))`, it would still get the error. Even though `CaseClassBar` is a case class, the way I make the function affects whether it works or not.

It would be nice if we can have the `layer1` style to work. That would definitely help us to migrate more easily.

```scala
trait Foo

  class Bar(a: String, b: Int) extends Foo {
    def doSomething(): String = a + b
  }

  object Bar {
    val layer1: ZLayer[String with Int, Nothing, Foo] = ZLayer.fromFunction(new Bar(_, _)) // doesn't work

    val layer2 = ZLayer.fromFunction(
      new Bar(_, _)
    ) // this works. And the layer can be used by  ZIO[Foo, _, _], as verified by below method

    def use(zio: ZIO[Foo, Nothing, Unit]): ZIO[Any, Nothing, Unit] =
      zio.provide(layer2, ZLayer.succeed(""), ZLayer.succeed(1))
  }

  case class CaseClassBar(a: String, b: Int) extends Foo {
    def doSomething(): String = a + b
  }

  object CaseClassBar {
    val layer3: ZLayer[String with Int, Nothing, Foo] = ZLayer.fromFunction(new CaseClassBar(_, _)) // doesn't work

    val layer4: ZLayer[String with Int, Nothing, Foo] = ZLayer.fromFunction(CaseClassBar(_, _)) // doesn't work

    val layer5: ZLayer[String with Int, Nothing, Foo] = ZLayer.fromFunction(CaseClassBar.apply _) // this works

    val layer6 = ZLayer.fromFunction(new CaseClassBar(_, _)) // this works

    val layer7 = ZLayer.fromFunction(CaseClassBar(_, _)) // this works

    val layer8 = ZLayer.fromFunction(CaseClassBar.apply _) // this works

    def use(zio: ZIO[Foo, Nothing, Unit]): ZIO[Any, Nothing, Unit] =
      zio.provide(layer8, ZLayer.succeed(""), ZLayer.succeed(1))
  }
```

**Comment by adamgfraser** [comment]
Yeah, unfortunately this runs into a Scala compiler type inference issue. One other way you can work around this is by assigning the result to an intermediate value.

```scala
val layer1: ZLayer[String with Int, Nothing, Foo] = {
  val layer = ZLayer.fromFunction(new Bar(_, _))
  layer
}
```

Unfortunately both the old `toLayer` syntax and `fromFunction` have warts in this regard. We have some ideas for an alternative but not sure what else we are going to be able to do here.

Copying @kitlangton.

**Comment by alphaho** [comment]
I see. Would it be possible to revisit the possibility of #6560? It would be great if we can have that implemented.

**Comment by kitlangton** [comment]
This is Scala 3, right?

**Comment by adamgfraser** [comment]
Yes that is the alternative. I think that may have tradeoffs of its own but we can certainly look at it.

**Comment by alphaho** [comment]
@kitlangton This is using Scala 2.13.6

**Comment by kitlangton** [comment]
That’s surprising! I’ll mess around later

**Comment by adamgfraser** [comment]
@alphaho Can you say more about why you are defining this as a class instead of a case class?

**Comment by alphaho** [comment]
@adamgfraser Sure. We tried the Service Pattern 2.0 before, by marking all the classes(both the data class and service class) as `case class` and using `(SomeClass(_, _, _)).toLayer` (we didn't know the `SomeClass.apply _` trick back then). 
But team members who are new to Scala found it quite hard to understand. 

Because they need to figure out 4 things at the same time:
1. the concept of ZLayer
2. why `apply` method generated for a case class
3. case class should usually be used for data classes, why this service class is marked as case class
4. how using the underscore would turn an expression to a lambda

In order to make it a bit easier to understand, we decided to drop the `case` modifier from the service class. And adding a `new` keyword doesn't feel like a lot of pain comparing to the problems it raises, especially we don't touch the layer definition a lot. Besides, having the `new` keyword can also help some new joiners find the way home and feel more familiar and less intimidating.
